src/Microsoft.ML.AutoML/API/ExperimentBase.cs

```diff
@@ -102,16 +102,17 @@ public abstract class ExperimentBase<TMetrics, TExperimentSettings>
             const int crossValRowCountThreshold = 15000;
 
             var rowCount = DatasetDimensionsUtil.CountRows(trainData, crossValRowCountThreshold);
+            bool isNotImageClassificationTask = columnInformation.ImagePathColumnNames.Count != 1;
 
-            if (rowCount < crossValRowCountThreshold)
+            if (rowCount < crossValRowCountThreshold && isNotImageClassificationTask)
```

Comment:
You'll need CrossVal when your count of images is low. Otherwise noise will dominate differences in your model metrics.

I assume ImageClassification will work in CV and you're just worried about runtime. Perhaps there's a more runtime sensitive heuristic that we could make.

Reply:
also, could you leave a comment about the rationale for the decision?

Reply:
Hey Justin, it is not about runtime, I just haven't seen anywhere on Tensorflow examples where they used CV for image classification and my other rational was that with CV is that it does not shuffle the data which means you could have all the images of cats, then all the images of dogs and with CV you might train on just cats and then test on dogs. I will try to add shuffling to CV as well so that it trains on all classes in every split.  

---
In reply to: [338960267](https://github.com/dotnet/machinelearning/pull/4384#discussion_r338960267) [](ancestors = 338960267)

Reply:
Shuffling is not required for CV to work properly. And CV is required when your number of rows is small, to get a useful measure of how good your model performs.

I think you're visualizing the CV folds as created as `folds =  [ rows 0 ... 99, rows 100...199, rows 200..299 ]`. Instead of this method, the rows are split pseudorandomly.

CV splits the dataset with each row going into one of K folds pseudorandomly. Then each of the K folds are chosen one at a time as the test dataset, and the other K-1 folds are used as the training set. 

The cats and dogs will be roughly equally split among the CV folds. That said, you may want a shuffle as part of your trainer, to lessen overfitting.

Reply:
Also, the `ShuffleRows()` does not do the expected style of shuffling. 

Most people would assume `ShuffleRows()` shuffles the dataset. Instead it does a block shuffle, where continuous segments of the dataset are shuffled only within that block/partition. Aka, rows at the top of the dataset will remain near the top; and rows at the bottom will remain near the bottom. Shuffling a dataset normally implies a row will have equal likelihood of ending anywhere.

The `ShuffleRows()` block shuffle allows for a streaming mode by taking only the block size (~1k rows) of memory at a time. There's no general way support full shuffling of dataset in a streaming pipeline.